doc/rbd/rbd-exclusive-locks.rst

```diff
@@ -0,0 +1,78 @@
+.. _rbd-exclusive-locks:
+
+====================
+ RBD Exclusive Locks
+====================
+
+.. index:: Ceph Block Device; RBD exclusive locks; exclusive-lock
+
+Exclusive locks are a mechanism designed to prevent multiple processes
+from accessing the same Rados Block Device (RBD) in an uncoordinated
+fashion. Exclusive locks are heavily used in virtualization (where
+they enable live migration, among other things), and also in RBD
+mirroring (where they are a prerequisite for journaling).
+
+Exclusive locks are only available in RBD format version 2. They are
+enabled by default, unless overridden via the ``rbd_default_features``
+configuration option or the ``--image-feature`` flag for ``rbd
+create``.
+
+In order to ensure proper exclusive locking operations, any client
+using an RBD image whose ``exclusive-lock`` feature is enabled should
+be using a CephX identity whose capabilities include ``profile rbd``.
+
+Exclusive locking is mostly transparent to the user.
+
+#. Whenever any ``librbd`` client process starts using an RBD image on
+   which exclusive locking has been enabled, it obtains an exclusive
+   lock on the image before the first write.
+
+#. Whenever any such client process gracefully terminates, it
+   automatically relinquishes the lock.
+
+#. This subsequently enables another process to acquire the lock, and
+   write to the image.
+
+Note that it is perfectly possible for two or more concurrently
+running processes to merely open the image, and also to read from
+it. The client acquires the exclusive lock only when attempting to
+write to the image.
+
+
+Blacklisting
+============
+
+When a process that previously held an exclusive lock on an image does
+not terminate gracefully, but is abruptly killed (due to having
+received a ``KILL`` or ``ABRT`` signal, for example, or a hard reboot
+or power failure), the exclusive lock is never gracefully
+released. Thus, when a new process starts and attempts to use the
+device, it needs a way to break the previously-held exclusive lock.
+
+However, a process may also hang, or merely lose network connectivity
+to the Ceph cluster for some amount of time. In that case, simply
+breaking the lock would be potentially catastrophic: the hung process
+or connectivity issue may resolve itself, and the old process may then
+compete with one that has started in the interim, accessing RBD data
+in an uncoordinated and destructive manner.
+
+Thus, in the event that a lock cannot be acquired in the standard
+graceful manner, the overtaking process does not only break the lock,
+but also blacklists the previous lock holder. This process is
+negotiated between the new process and the Ceph Mon: upon receiving
+the exclusive lock and blacklist request,
+
+* the Mon instructs the relevant OSDs to no longer serve requests from
+  the old process;
+* once the OSDs confirm, the Mon grants the lock to the new process;
```

Comment:
The blacklist is distributed in a lazy fashion from the MON to the OSDs via regular OSD map updates. There is no explicit blacklist communication between MONs and OSDs. The new client that issued the blacklist waits for this OSD map update so that all future issued IOs will include the required OSD map epoch -- which would ensure the OSDs force-refresh their OSD map if needed prior to processing the incoming IO.

Reply:
Ack. Replacing "once the OSDs confirm" with "once the associated OSD map update is complete".